Function to create age groups

chrisdeans opened the pull request.

jackhannah95 commented:
Hi @chrisdeans,

Good work 👍 this will be a good addition to the package.

I haven't looked at the tests or documentation much yet, because I think the function(s) should be restructured slightly, and that will presumably have a knock-on effect.

Overall I think only one of the functions needs to exist, and I think it should be `get_agegroup_seq()` (although I would just call it `age_group()`; the "get" seems a bit superfluous to me). I think that the `breaks` argument in `get_agegroup()` will be quite hard to understand for less-experienced R users (which is probably why everyone uses a giant `case_when()` instead of `cut()` as it is).

I think users experienced enough to understand `breaks = c(seq(0, 90, by = 5), Inf)` are probably also experienced enough to be able to bypass this function entirely and use

```r
cut(ages, 
    breaks = c(seq(0, 90, by = 5), Inf),
    labels = c(paste(seq(0, 85, by = 5), 
                     seq(4, 84, by = 5),
                     sep = "-"), 
               paste(90, "+", sep = "")))
```

to get the same result. `get_agegroup_seq()` seems much more user-friendly and, ultimately, beginner-friendly to me, so I'd be inclined to go with that.

I think there are probably also too many arguments in `get_agegroup()`. If `get_agegroup_seq()` becomes the overarching `age_group()` function then I don't think it needs the `include_zero`, `include_inf`, `sep` or `above.char` arguments. Adding that many arguments increases the function's complexity and I don't think it needs to be quite as flexible. Anyone wishing to exclude certain ages from this will presumably have filtered them out already, and I really think that anyone not happy with a `-` separating numbers or a `+` signifying "and over" can lump it. They're not hard to change afterwards if anyone is so inclined.

I've hummed and hawed about the `as_factor` argument vs just making it a character by default, but I suppose the factor setting is useful in a `group_by()` with `.drop = FALSE` so I'd be happy for that to stay.

There's quite a lot of error handling in the `get_agegroup()` that could hopefully be simplified a little by combining into one function and removing the `include_zero` and `include_inf` arguments. I'll leave looking too closely at those for now.

I hope these comments seem reasonable. I'm finishing at ISD on April 17th and ideally I'd like to do another release of the package with this and another couple of functions included before I go. Hopefully that's doable but not the end of the world if you don't get around to it.

Cheers,

Jack

chrisdeans commented:
Hi @jackhannah95 ,

I've updated the function (now `age_group()`) along the lines you suggested. Could you have another look over it at some point?

It turns out after you take out those parameters all of the error handling also disappears; `cut()`, `seq()` and `as.numeric()` can handle any remaining error cases. 

I've left in `as_factor` for the moment , but I've changed the default to `FALSE` i.e. character.

Thanks,
Chris

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+#'
+#' @description
+#' A wrapper for \code{\link[base]{cut}} tailored for creating age groups. For
+#' each value of integer \code{x}, what age group does it fall into.
```
I think I would re-word this slightly to something like "An extension of `cut()`. `age_group()` takes a numeric vector and assigns each age to the appropriate age group."

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+#'
+#' @param x a vector of numeric values
+#' @param from,to,by start, end and increment values for a sequence of age
+#'   groups.
```
I think I would do a separate param for each of these and explain what the default values are.

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+#'   \code{TRUE} to return a factor vector instead.
+#'
+#' @return A character vector, where each element is the age group for the
+#'   corresponding element in \code{x}. If \code{as_factor = FALSE} a factor
```
I think this should be "if `as_factor = TRUE`, a factor is returned instead".

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+#' The \code{from},\code{to} and \code{by} are used to create distinct age
+#' groups \code{from-from+by}, ..., up to \code{to+}, or a value less than
+#' \code{to} if \code{to} is not an exact multiple. The dafault values provide
+#' the standard five year age groups.
```
This section is a little difficult to follow. I think I would re-word it to something like:

"The `from`, `to` and `by` values are used to create distinct age groups. `from` dictates the starting age of the lowest age group, and `by` indicates how wide each group should be. `to` stipulates the cut-off point at which all ages equal to or greater than this value should be categorised together in a `to+` group. If the specified value of `to` is not a multiple of `by` then the value of `to` is rounded down to the nearest multiple of `by`. The default values of `from`, `to` and `by` are those which split ages into the [European Standard Population](https://www.opendata.nhs.scot/dataset/standard-populations/resource/edee9731-daf7-4e0d-b525-e4c1469b8f69) age groups."

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+#' \code{to} if \code{to} is not an exact multiple. The dafault values provide
+#' the standard five year age groups.
+#'
+#' @export
```
This is pernickety but I would move export to after the examples.

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+#' age_group(age, 0, 65, 10)
+#'
+#' #To get the output as a factor:
+#' age_group(age, as_factor = TRUE)
```
I think I would include the argument names in the examples (apart from `x = age`) so it's absolutely clear what number relates to what argument.

Also, very pernickety but if you wouldn't mind putting a space after a `#` it would be much appreciated.

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+#'
+
+age_group <- function(x, from = 0, to = 90, by = 5,
+                      as_factor = FALSE) {
```
Also pernickety but if the arguments are going over multiple lines then could they each go on their own line?

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+  # Reformat label for last value
+  labels <- gsub("-Inf", "+", labels)
+
+  agegroup <- cut(as.numeric(x),
```
I've gone back and forth on the `as.numeric()` here a lot and I think it would be safer to take it out and let `cut()` error if it's not supplied with a numeric. I know it's convenient but I think silent type conversion is a bit risky if someone doesn't know it's happening.

jackhannah95 commented on R/age_group.R:
```diff
@@ -0,0 +1,60 @@
+  if(as_factor == F)
+    agegroup <- as.character(agegroup)
+
+  return(agegroup)
```
The `return()` is probably superfluous here; you could just type `agegroup` and it'd produce the same output.

jackhannah95 commented on tests/testthat/test-age_group.R:
```diff
@@ -0,0 +1,53 @@
+
+  expect_warning(a <- age_group(c("1", "57", "apple", "12"), as_factor = F))
+  expect_equal(a, c("0-4", "55-59", NA, "10-14"))
+})
```
Overall the tests are absolutely great, really thorough. I just wondered though if a comment could be added above the `expect_warning()` ones (or `expect_error()` if they change) to explain why a warning/error is expected for anyone looking at a glance (I realise this is a bit hypocritical because I mostly don't do this on my own `expect_warning()`  and `expect_error()` tests even though I should).

I guess removing `as.numeric()` from the function might alter these slightly but, assuming nothing would change for the sake of argument, I would define `a` outside the `expect_*` and maybe call the second vector `b` instead of overwriting `a` just so that flipping the order around won't affect the results of the tests.

chrisdeans commented:
Hi @jackhannah95, thanks for the in-depth comments! I've added a new version, hopefully covering everything you mentioned. Let me know if you spot anything else!
